solr/core/src/java/org/apache/solr/cluster/placement/plugins/AffinityPlacementFactory.java

```diff
@@ -280,25 +250,31 @@ public void verifyAllowedModification(ModificationRequest modificationRequest, P
 
     private void verifyDeleteCollection(DeleteCollectionRequest deleteCollectionRequest, PlacementContext placementContext) throws PlacementModificationException, InterruptedException {
       Cluster cluster = placementContext.getCluster();
-      Set<String> colocatedCollections = colocatedWith.getOrDefault(deleteCollectionRequest.getCollection().getName(), Set.of());
-      for (String primaryName : colocatedCollections) {
-        try {
-          if (cluster.getCollection(primaryName) != null) {
-            // still exists
-            throw new PlacementModificationException("colocated collection " + primaryName +
-                " of " + deleteCollectionRequest.getCollection().getName() + " still present");
-          }
-        } catch (IOException e) {
-          throw new PlacementModificationException("failed to retrieve colocated collection information", e);
+      Set<String> colocatedWith = new HashSet<>();
+      for (SolrCollection collection : cluster.collections()) {
```

Comment:
This loop is frightening... Looping over all collection will be expensive. Can't we instead update collection properties with the "reverse" of the `withCollection` relation?
Or that means that for high scale SolrCloud clusters (WIP), `withCollection` is strongly discouraged.

Reply:
That's the most serious downside of keeping this config in collection properties - there's no easy way to keep the reverse mapping up to date. We could try updating the secondary collection... although this would require adding special case code to `OCMH.modifyCollection()`, which breaks the isolation of the plugin from the core.

Reply:
We could document to the user that in order to support safe deletes, the secondary collection must reference the primary one. Then up to the user to decide if the two collections are updated or not...
At least this leaves a scalable option available.
The other "cleaner" solution is to make collection references a first class citizen managed correctly (so we can maintain mapping and reverse mapping), but then the scope increases greatly and changes quite a few things in how collection are manages (and `withCollection` would then be just one specific usage of this collection reference mechanism, but maybe the only one...).

No perfect solution here, but once we get to very large numbers of collections (with collection lists not stored in node memory) the iteration will be prohibitively expensive (that's why I didn't want having such an iterator in the first place BTW).

Reply:
Hmm, then at this point it makes more sense to keep this configuration in the plugin config :(

Reply:
I disagree, but your call.